ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java

```diff
@@ -479,7 +485,12 @@ private void readImpl(DataStreamRequestByteBuf request, ChannelHandlerContext ct
       try {
         if (exception != null) {
           replyDataStreamException(server, exception, info.getRequest(), request, ctx);
-          removeDataStream(key, info);
+        }
+        if (close || exception != null) {
```

Comment:
For normal close case,
- the local stream (to a local file) is a `DataStream` created by an application.  It should be  removed from the `DataStreamMap` and closed during `link(..)`.
- Remote streams (server to server) should be closed here.  This is the leak.

To play safe, call `info.cleanUp(key)` when `removed == null`.  The code becomes
```java
        if (exception != null) {
          replyDataStreamException(server, exception, info.getRequest(), request, ctx);
          final StreamInfo removed = removeDataStream(key);
          if (removed != null) {
            Preconditions.assertSame(info, removed, "removed");
          } else {
            info.cleanUp(key);
          }
        } else if (close) {
          info.applyToRemotes(remote -> remote.out.closeAsync());
        }
```

Reply:
```
 else if (close) {
          info.applyToRemotes(remote -> remote.out.closeAsync());
        }
```
Do we need this? remotes should have done this in remote write.

Reply:
Could you show the code?  I cannot find the code closing the remote streams.

Reply:
I think for close request, remote has already write here? https://github.com/apache/ratis/blob/master/ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java#L463

Reply:
It will call `DataStreamOutputImpl.writeAsync` but not `DataStreamOutputImpl.close()`.

Reply:
If it's a close message, writeAsync will send the close message to the next peer.

`info.applyToRemotes(remote -> remote.out.closeAsync());` will send an empty close message again.